### ISSUE TITLE
Active Storage's secret is not stable and causes cached proxy urls to break after a while

### ISSUE THREAD

**Issue by brenogazzola** [rails/rails#40614]
### Steps to reproduce
Haven't found a way to reliably reproduce this one, but what usually happens is:

1 - Use fragment caching to cache the proxy url of an S3 image;
2 - Wait a few days/deploys. The image will eventually break
3 - Compare the cached url, with a new version of the same url. They will be different.

As an example, here's a the signed id portion of a proxy url I currently have cached:
```
eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOGo4RFE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--6077681b4217971fe39b06ad8079637bbbb8a977
```

And here is what it looks like now:
```
eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOGo4RFE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--7d5e1ca15aa8e20630757bf8df291a1d17713d81
```

As you can see the part generated by the message verifier using the `id` and the `:blob_id` as purpose remains the same. But the second part (the part after `--`) which is the digest has changed.

The digest is generated by the `generate_digest` which uses the first part and a `@secret` variable, suggesting that the secret is changing for some reason.

@georgeclaghorn 
@fleck 

### Expected behavior
Proxy urls should remain the same

### Actual behavior
Proxy URLs are changing

### System configuration
**Rails version**: 53331b478d32e42a8155414379b2edbe4277a3a4
**Ruby version**: 2.7.2

**Comment by composerinteralia** [comment]
Any chance your `secret_key_base` is getting changed? I poked around in the code and couldn't think of anything else that would cause this. 

I'll leave the files I looked through here in case it helps you debug:

The signed_id comes from this method:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/activestorage/app/models/active_storage/blob.rb#L146-L148

Which calls `super` to the Active Record method:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/activerecord/lib/active_record/signed_id.rb#L112-L114

`self.class.signed_id_verifier` in this case is:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/activestorage/app/models/active_storage/blob.rb#L140-L142

which gets set in engine.rb:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/activestorage/lib/active_storage/engine.rb#L111

It is a `MessageVerifier` with a secret passed in on initialization:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/railties/lib/rails/application.rb#L199-L204

That secret is defined via a `KeyGenerator`:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/railties/lib/rails/application.rb#L173-L179

Which initializes with the secret key base:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/railties/lib/rails/application.rb#L426-L434

The secret key comes from calling `generate_key` on the key generator. With the same secret (i.e. the secret key base) and salt (i.e. "ActiveStorage"), this should always return the same generated key:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/activesupport/lib/active_support/key_generator.rb#L22-L24

That generated key gets passed to the `MessageVerifier` and assigned to the instance variable `@secret` that you mentioned. That gets used in `generate_digest`:

https://github.com/rails/rails/blob/53331b478d32e42a8155414379b2edbe4277a3a4/activesupport/lib/active_support/message_verifier.rb#L200-L203

In short, I think the digest is roughly:

```rb
OpenSSL::HMAC.hexdigest(
  OpenSSL::Digest::SHA1.new,
  OpenSSL::PKCS5.pbkdf2_hmac_sha1("secret key base", "ActiveStorage", 1, 64),
  "data"
)
```

So if that is changing, I am thinking that the secret key base must be changing.

**Comment by composerinteralia** [comment]
Also noticed https://github.com/rails/rails/pull/39623, which you might be interested in if you are indeed rotating your secret key base.

**Comment by brenogazzola** [comment]
Unfortunately no. Both the `secret_key_base` which is in my credentials file and the `RAILS_MASTER_KEY` which is used to decrypt it have not been changed since the app was first created.

The first part of the url is generated by the `message_verifier`, so I believe that is the `secret_key_base` was had changed, that part would have changed too. But only the `digest` was affected, and this is what making me confused.

**Comment by santib** [comment]
> Unfortunately no. Both the `secret_key_base` which is in my credentials file and the `RAILS_MASTER_KEY` which is used to decrypt it have not been changed since the app was first created.
> 
> The first part of the url is generated by the `message_verifier`, so I believe that is the `secret_key_base` was had changed, that part would have changed too. But only the `digest` was affected, and this is what making me confused.

Is it possible that - even if the `secret_key_base` in your credentials file didn't change - a different secret key base is getting retrieved? For example if someone set the `ENV['SECRET_KEY_BASE']` it will have priority over other ways of setting it..

```ruby
    # The secret_key_base is used as the input secret to the application's key generator, which in turn
    # is used to create all MessageVerifiers/MessageEncryptors, including the ones that sign and encrypt cookies.
    #
    # In development and test, this is randomly generated and stored in a
    # temporary file in <tt>tmp/development_secret.txt</tt>.
    #
    # In all other environments, we look for it first in ENV["SECRET_KEY_BASE"],
    # then credentials.secret_key_base, and finally secrets.secret_key_base. For most applications,
    # the correct place to store it is in the encrypted credentials file.
    def secret_key_base
      if Rails.env.development? || Rails.env.test?
        secrets.secret_key_base ||= generate_development_secret
      else
        validate_secret_key_base(
          ENV["SECRET_KEY_BASE"] || credentials.secret_key_base || secrets.secret_key_base
        )
      end
    end
```

**Comment by brenogazzola** [comment]
You were right. My credentials say `4c4af`, but the command you gave me returns `5df55e`. I've checked my settings on Heroku but its not there. However a `printenv` revealed it. Guess I have to take this to the heroku team. Thanks a lot @santib
